README.md
# Flutter FAQs [Frequently Asked Question]

This plugin helps you to display FAQs in your Flutter projects as a custom widget.

## Table of contents

- [Features](#features)
- [Installation](#installation)
- [Usages](#usages)
- [Support and feedback](#support-and-feedback)


## Features

You can:
*  use this for showing FAQs.
*  add textstyle for both questions and answers.
*  give your icons on expnding and collasped.
*  give padding to answers.
*  show divider or not or your own widget as a separator.


## Installation

Add package to the the pubspec.yaml.

```dart
flutter_faq:
```

## Usages

Import the package.

```dart
import 'package:flutter_faq/flutter_faq.dart';
```

Add FAQ widget to the widget tree

```dart
FAQ(question: "Question 1", answer: "data"),
```

For Styling Questions and Answers

```dart
FAQ(
    question: "Question",
    answer: "data",
    ansStyle: const TextStyle(color: Colors.blue, fontSize: 15),
    queStyle: const TextStyle(color: Colors.green, fontSize: 35),
    ),
```

For giving color and borderRadius

```dart
FAQ(
    question: "Question",
    answer: data,
    ansDecoration: BoxDecoration(
        color: Colors.grey[550],
        borderRadius: const BorderRadius.all(Radius.circular(20))),
    queDecoration: BoxDecoration(
        color: Colors.grey[500],
        borderRadius: const BorderRadius.all(Radius.circular(20))),
    ),
```

For giving padding in answer

```dart
FAQ(
    question: "Question",
    answer: "data",
    ansPadding: const EdgeInsets.all(50),
    ),
```

For giving custom divider

```dart
FAQ(
    question: "Question",
    answer: "data",
    separator: Container(
        height: 5,
        width: double.infinity,
        color: Colors.purple,
    ),
    ),
```

For giving custom expanded and collapsed icon

```dart
FAQ(
    question: "Question",
    answer: "data",
    expandedIcon: const Icon(Icons.minimize),
    collapsedIcon: const Icon(Icons.add),
    ),
```

For no showing divider

```dart
FAQ(
    question: "Question",
    answer: "data",
    separator:false
    ),
```


## Support and feedback

If I miss to add some feature, feel free to open a ticket or contribute!
You can reach me here:
[LinkedIn](https://www.linkedin.com/in/lakshydeep-14/) | 
[GitHub](https://github.com/lakshydeep-14) | 
[Medium](https://lakshydeep-14.medium.com/) | 
[Google DevLibrary](https://devlibrary.withgoogle.com/authors/lakshydeep-14)

CHANGELOG.md
## 1.0.1

* width added

## 1.0.1

* Tile color and border radius added

## 1.0.0

* TODO: Describe initial release.

lib/flutter_faq.dart
library flutter_faq;

import 'package:flutter/material.dart';

class FAQ extends StatefulWidget {
  // for question
  final String question;

  //answer
  final String answer;

  //question textstyle, defaults to EdgeInsets.symmetric(horizontal: 35.0, vertical: 10),
  final TextStyle? queStyle;

  // answertextstyle, defaults to  TextStyle(color: Colors.black,fontWeight: FontWeight.w500,fontSize: 18),
  final TextStyle? ansStyle;

  //defaults to false
  final bool? isExpanded;

  //defaults to Icon.chevron_right
  final Widget? expandedIcon;
  final Widget? collapsedIcon;

  //defaults to true
  final bool showDivider;

  //padding to give in answer
  final EdgeInsets? ansPadding;

  //divider
  final Widget? separator;

  final BoxDecoration? ansDecoration, queDecoration;

  const FAQ(
      {super.key,
      required this.question,
      required this.answer,
      this.queStyle,
      this.ansStyle,
      this.isExpanded = false,
      this.expandedIcon,
      this.collapsedIcon,
      this.showDivider = true,
      this.ansPadding,
      this.separator,
      this.ansDecoration,
      this.queDecoration});

  @override
  State<FAQ> createState() => _FAQState();
}

class _FAQState extends State<FAQ> {
  @override
  Widget build(BuildContext context) {
    return ListExpandableWidget(
      ansDecoration: widget.ansDecoration ??
          BoxDecoration(
              color: Colors.grey[200],
              borderRadius: const BorderRadius.all(Radius.circular(20))),
      queDecoration: widget.queDecoration ??
          BoxDecoration(
              color: Colors.grey[200],
              borderRadius: const BorderRadius.all(Radius.circular(20))),
      ansPadding: widget.ansPadding ??
          const EdgeInsets.symmetric(horizontal: 35.0, vertical: 10),
      separator: widget.separator ?? const Divider(),
      showDivider: widget.showDivider,
      isExpanded: widget.isExpanded,
      expandedIcon:
          widget.expandedIcon ?? const Icon(Icons.keyboard_arrow_down),
      collapsedIcon:
          widget.collapsedIcon ?? const Icon(Icons.keyboard_arrow_right),
      header: Text(
        widget.question,
        style: widget.queStyle ??
            const TextStyle(
                color: Colors.red, fontWeight: FontWeight.bold, fontSize: 20),
      ),
      items: Text(
        widget.answer,
        textAlign: TextAlign.justify,
        style: widget.ansStyle ??
            const TextStyle(
                color: Colors.black, fontWeight: FontWeight.w500, fontSize: 18),
      ),
    );
  }
}

class ListExpandableWidget extends StatefulWidget {
  // optional property and default value is false
  final bool? isExpanded;

  final Widget separator;

  // required widget and display the header of each group
  final Widget? header;

  // required list `ListTile` widget for group items
  final Widget items;

  // optional widget for expanded Icon. Default value is `Icon(Icons.keyboard_arrow_down)`
  final Widget expandedIcon;

  final EdgeInsets ansPadding;
  final bool showDivider;
  final BoxDecoration ansDecoration, queDecoration;
  // optional widget for collapse Icon. Default value is `Icon(Icons.keyboard_arrow_right)`
  final Widget collapsedIcon;
  const ListExpandableWidget(
      {super.key,
      this.isExpanded,
      this.header,
      required this.showDivider,
      required this.items,
      required this.expandedIcon,
      required this.collapsedIcon,
      required this.separator,
      required this.ansPadding,
      required this.ansDecoration,
      required this.queDecoration});

  @override
  State<ListExpandableWidget> createState() => _ListExpandableWidgetState();
}

class _ListExpandableWidgetState extends State<ListExpandableWidget> {
  bool? _isExpanded;

  @override
  void initState() {
    super.initState();
    _updateExpandState(widget.isExpanded!);
  }

  void _updateExpandState(bool isExpanded) =>
      setState(() => _isExpanded = isExpanded);

  @override
  Widget build(BuildContext context) {
    return _isExpanded! ? _buildListItems(context) : _wrapHeader();
  }

  Widget _wrapHeader() {
    List<Widget> children = [];
    if (!widget.isExpanded!) {
      children.add(widget.showDivider == false
          ? const SizedBox(
              height: 5,
            )
          : widget.separator);
    }
    children.add(Container(
      decoration: widget.queDecoration,
      child: ListTile(
        style: ListTileStyle.drawer,
        title: widget.header,
        trailing: _isExpanded! ? widget.expandedIcon : widget.collapsedIcon,
        onTap: () => _updateExpandState(!_isExpanded!),
      ),
    ));
    return Ink(
      child: Column(
        children: children,
      ),
    );
  }

  Widget _buildListItems(BuildContext context) {
    List<Widget> titles = [];
    titles.add(_wrapHeader());
    titles.add(Container(
        width: double.infinity,
        decoration: widget.ansDecoration,
        padding: widget.ansPadding,
        child: widget.items));
    return Column(
      crossAxisAlignment: CrossAxisAlignment.center,
      children: ListTile.divideTiles(tiles: titles, context: context).toList(),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_faq/flutter_faq.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        body: SingleChildScrollView(
          padding: const EdgeInsets.symmetric(horizontal: 20),
          child: Column(children: [
            FAQ(
              question: "Question 1",
              answer: data,
              ansDecoration: BoxDecoration(
                  color: Colors.grey[500],
                  borderRadius: const BorderRadius.all(Radius.circular(20))),
              queDecoration: BoxDecoration(
                  color: Colors.grey[500],
                  borderRadius: const BorderRadius.all(Radius.circular(20))),
            ),
            FAQ(question: "Question2", answer: data),
            FAQ(
              question: "Question 3",
              answer: data,
              ansStyle: const TextStyle(color: Colors.blue, fontSize: 15),
              queStyle: const TextStyle(color: Colors.green, fontSize: 35),
            ),
            FAQ(
              question: "Question 4",
              answer: data,
              showDivider: false,
            ),
            FAQ(
              question: "Question 5",
              answer: data,
              expandedIcon: const Icon(Icons.minimize),
              collapsedIcon: const Icon(Icons.add),
              showDivider: false,
              ansStyle: const TextStyle(color: Colors.blue, fontSize: 15),
              ansPadding: const EdgeInsets.all(50),
            ),
            FAQ(
              question: "Question 6",
              answer: data,
              expandedIcon: const Icon(Icons.minimize),
              collapsedIcon: const Icon(Icons.add),
              ansStyle: const TextStyle(color: Colors.blue, fontSize: 15),
              ansPadding: const EdgeInsets.all(50),
              separator: Container(
                height: 5,
                width: double.infinity,
                color: Colors.purple,
              ),
            ),
            FAQ(
              question: "Question 7",
              answer: data,
              expandedIcon: const Icon(Icons.minimize),
              collapsedIcon: const Icon(Icons.add),
              showDivider: false,
              ansStyle: const TextStyle(color: Colors.blue, fontSize: 15),
              ansPadding: const EdgeInsets.all(50),
            ),
          ]),
        ),
      ),
    );
  }
}

String data = """"
Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
""";
